### ISSUE TITLE
trailers are not being passed correctly in the response

### ISSUE THREAD

**Issue by rauanmayemir** [connectrpc/vanguard-go#139]
Trailers are being broken in the response when grpc is served behind vanguard:
<img width="518" alt="image" src="https://github.com/user-attachments/assets/d4b848b6-b534-447b-9164-64df11e858db">

I'm assuming that under the hood vanguard essentially wraps `grpcServer.ServeHttp` which leads to the same outcome when used as is without vanguard. There is a similar issue that was never really solved: https://github.com/grpc/grpc-go/issues/2206.

**Comment by emcfarlane** [comment]
@rauanmayemir thanks for raising the issue. I'm going to add the [conformance test suite ](https://github.com/connectrpc/vanguard-go/issues/141) to improve coverage which will hopefully highlight this issue and allow us to investigate solutions. Issues https://github.com/grpc/grpc-go/issues/2206 and https://github.com/grpc/grpc-go/issues/553 look like the cause so ideally we could upstream a fix if we are unable to workaround it.

**Comment by rauanmayemir** [comment]
Wondering if there's any chance to have this fixed soon.

I've been holding off on adopting vanguard due to this issue breaking my frontend that depends on gRPC-Web. To understand response state we need to parse `grpc-status` and `grpc-status-details-bin` headers which are not available especially when it's a non-OK response.

(I could take a stab and try fixing this, but I don't know where to start)

**Comment by jhump** [comment]
@rauanmayemir, I started looking into this and am trying to figure out what _exactly_ the issue is. The screenshot above and the linked issue don't really describe an error. Is it that application-set trailers are missing from the response? Or are the actual status and details failing to come across (like an RPC that should fail is being incorrectly considered a success)?

The linked issue just describes that the handler never emits trailers-only responses. But such responses are not necessarily expected/required by the spec -- separating them into headers and trailers is acceptable (and, in fact, they _must_ be separated if the application has set any custom headers since metadata in a trailers-only response are interpreted as trailers).

I'll keep playing with this to see if I can reproduce some sort of problem, but I think I'm uncertain from the above on exactly what the expected behavior is and what's going wrong.

**Comment by rauanmayemir** [comment]
Thank you, I’ll try to reproduce and find the issue as well. The issue here is that headers are being garbled like in the screenshot (it’s from grpcui, I’ll try to find out what’s actually happening).

**Comment by rauanmayemir** [comment]
@jhump It's the obvious part here:

```
trailer: Grpc-Status
trailer: Grpc-Message
trailer: Grpc-Message-Details-Bin
```

What I should have seen is:

```
Grpc-Status: 13
Grpc-Message: Error message
Grpc-Message-Details-Bin: base64_error_details_payload
```

Also notice `Accept-Encoding`.

**Comment by jhump** [comment]
> What I should have seen is:

@rauanmayemir, the `Grpc-Status` etc will be trailers, not headers.

So is the issue that the `grpcui` reported the RPC as _successful_, instead of failing with an "internal" error code?

> Also notice `Accept-Encoding`.

That looks fine. Are the `Vary` headers posing an issue for you?

**Comment by rauanmayemir** [comment]
If I remember correctly, it will interpret rpc as failed.

I think I'm confusing grpcui's visual presentation with grpc-web/grpc-json transcoding of headers and trailers into http headers. But it still bothers me that some requests correctly show my trailers and headers separately (i.e: headers, body, trailers), but for unsuccessful requests with `ServeHttp` trailers and headers turn into a mess.

Let me get back to you after I try grpc-web/grpc-json requests instead of grpcui.

**Comment by jhump** [comment]
> but for unsuccessful requests with ServeHttp trailers and headers turn into a mess.

Do you mind including a more full screenshot to demonstrate what you mean? The one above seems to be clipped and only shows the headers. There should be more, like an error and optional trailers, below that. This could also potentially be a bug/quirk with how `grpcui` renders these responses.

**Comment by rauanmayemir** [comment]
That’s what I meant, there is only what is displayed in the screenshot. Since this is an unsuccessful response, there is no body. I have no idea what is headers and what is trailers, all of it is in there together.

> This could also potentially be a bug/quirk with how grpcui renders these responses.

Very possible, which is why I’m coming up with a repro so the metadata can be textual.

**Comment by rauanmayemir** [comment]
There is indeed an issue with the trailers, they are being included to response headers in a weird way.

Here's an output from grpcui:
<img width="529" alt="image" src="https://github.com/user-attachments/assets/db4aec42-4399-4b55-83ef-8ee7e12b6d90" />

And here's a variant with `grpcurl`:

```
grpcurl -d '{}' -plaintext -vv 127.0.0.1:35071 my.package.ServiceName/MethodName

Resolved method descriptor: [skipped]

Response headers received:
content-type: application/grpc
trailer: Grpc-Status
trailer: Grpc-Message
trailer: Grpc-Status-Details-Bin

Response trailers received:
(empty)
Sent 1 request and received 0 responses
ERROR:
  Code: FailedPrecondition
  Message: Request validation error
  Details:
  1)	{"@error":"google.rpc.ErrorInfo is not recognized; see @value for raw binary message data","@type":"type.googleapis.com/google.rpc.ErrorInfo","@value":"payload"}
  2)	{"@error":"google.rpc.BadRequest is not recognized; see @value for raw binary message data","@type":"type.googleapis.com/google.rpc.BadRequest","@value":"payload"}
```

@jhump Btw, it seems like vanguard is actually the one adding those headers here: https://github.com/connectrpc/vanguard-go/blob/9c8fc536b41272f368cfa57ce7ff91915af9123f/protocol_grpc.go#L69

It is unclear to me why is it done.

**Comment by rauanmayemir** [comment]
Tried for `grpc-web:

```
curl -vv -X POST "http://127.0.0.1:35071/my.package.ServiceName/MethodName" -H 'x-grpc-web: 1' -H 'Content-Type: application/grpc-web+proto' -H 'Accept: application/grpc-web+proto' --data-binary @req.bin --output -
Note: Unnecessary use of -X or --request, POST is already inferred.
*   Trying 127.0.0.1:35071...
* Connected to 127.0.0.1 (127.0.0.1) port 35071
> POST /my.package.ServiceName/MethodName HTTP/1.1
> Host: 127.0.0.1:35071
> User-Agent: curl/8.7.1
> x-grpc-web: 1
> Content-Type: application/grpc-web+proto
> Accept: application/grpc-web+proto
> Content-Length: 11
> 
* upload completely sent off: 11 bytes
< HTTP/1.1 200 OK
< Content-Type: application/grpc-web+proto
< Transfer-Encoding: chunked
< 
p
%%%PAYLOAD%%% Grpc-Message: 
Grpc-Status: 0
* Connection #0 to host 127.0.0.1 left intact
```

I'm guessing `p` is misinterpreted 4 bytes that contain response data length.

As you can see here, `Grpc-Message` and `Grpc-Status` trailers aren't 'hoisted' as a response header. They should have been, envoy converts them to response headers.

**Comment by jhump** [comment]
@rauanmayemir, this all looks fine.

The first example isn't "hoisting" anything into headers. It is simply pre-declaring some of the trailer fields. This is an implementation detail of Go's ["net/http"](https://pkg.go.dev/net/http) package: in order to make sure the package will send trailers separately from headers when there is no response body, one must predeclare a trailer.

That "trailer" response header is part of the HTTP standard: https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Trailer

You will not see this with the grpc-web protocol because grpc-web does not actually use trailers: the trailer metadata and status are included in a special message at the end of the response body (because browsers APIs generally don't support trailers, so the only way for grpc-web browser JavaScript client code to observe the status is to put it in the body).

The reason you may not see this with other gRPC server implementations is that they don't use the "net/http" library, so don't have the same caveat imposed to make sure trailers are correctly sent. In some cases, the "net/http"-based implementation _could_ send what's called a "trailers-only" response -- in this case, the response headers contain any trailing metadata and the status keys (designed to make the response more compact over HTTP/2 when there is no response body data and the application isn't trying to send custom headers separately from custom trailers).

Vanguard uses "net/http", and then requires the use of `grpc.Server.ServeHTTP`, which is an adapter for `*grpc.Server` instances to work with the "net/http" package, and that's why you see this behavior with Vanguard, but not when using the grpc protocol directly to the `grpc.Server` if using its normal `Serve` method.

**Comment by rauanmayemir** [comment]
> the trailer metadata and status are included in a special message at the end of the response body

Oh, you mean I was supposed to parse trailers from the response body in the first place and not rely on http headers.

I checked the [protocol](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-WEB.md#protocol-differences-vs-grpc-over-http2) and you are right. I thought it was a bug in grpc-go when in fact it's a 'bug' in envoy: for trailer-only responses it hoists `grpc-message`, `grpc-status`, `grpc-status-details-bin` to headers and just return empty response instead.

Should vanguard follow the protocol then? It shouldn't be very hard to parse trailers from the body and as a bonus, it will support larger values.

**Comment by rauanmayemir** [comment]
OTOH, [browser features](https://github.com/grpc/grpc-web/blob/master/doc/browser-features.md#cors-support) section for CORS mentions `grpc-status` and `grpc-message` headers explicitly.

**Comment by rauanmayemir** [comment]
I'm getting really confused. In case of non-OK responses vanguard _DOES NOT_ fail, it will indeed add `grpc-*` headers and return empty body. (I can't remember why I couldn't reproduce this before)

The only instance where it differs from envoy is when grpc-status is 0. Envoy will still add it, but vanguard won't.

Should we close this issue? 😄

**Comment by jhump** [comment]
> Should we close this issue? 😄

@rauanmayemir, I think so. If you client can access application-specified headers and trailers and error statuses sent from the server aren't being incorrectly transformed into "OK" statuses (or vice versa), then I don't think there's any bug here. While the responses from vanguard may look a little different from other gRPC server implementations, they are still valid.

**Comment by rauanmayemir** [comment]
Okay, something's definitely not right. Caught an issue where error response trailers were not hoisted into response headers:

<img width="396" alt="image" src="https://github.com/user-attachments/assets/0481aea3-79ef-4f92-ae45-a7ba4c9bba6f" />

And response headers are:

<img width="531" alt="image" src="https://github.com/user-attachments/assets/492753e9-f36d-46bf-99c0-64376e02325e" />

I'll add an extra step and check trailers inside payload, but it's odd how this changes from build to build.

**Comment by jhump** [comment]
@rauanmayemir, I still don't understand why the above is a problem. When you say "hoisted into response headers", do you mean a "trailers-only response"?

Also, are you running into actual _semantic_ errors between clients and servers? Or are you just examining things on-the-wire and seeing things you don't expect? If the latter, I don't think that's necessarily a bug. Describing actual communication problems between clients and servers, like erroneous client-side behavior due the presence of vanguard, would be much more helpful for determining if this is a bug and, if so, how to correct it.

**Comment by rauanmayemir** [comment]
The problem is that I can’t have reproducible behaviour where non-OK trailer-only responses can be converted to empty bodies with trailers being hoisted into response headers. Same build can lead to different outcome depending on unknown to me yet conditions.

**Comment by jhump** [comment]
> The problem is that I can’t have reproducible behaviour where non-OK trailer-only responses can be converted to empty bodies with trailers being hoisted into response headers.

This is still very unclear. I asked before if you meant a "trailers-only" response when you say "hoisted into response headers". However, the above suggests that's _not_ what you mean. Are you meaning the "Trailer" header keys, that predeclare the trailers? I would love if we could reach consensus on language that makes it clear to botgh of us that we understand one another and aren't inadvertently talking past each other.

Also, what exactly are you building? Are you trying to write some sort of low-level interceptor or something that is interpreting the wire format? How do actual _clients_ behave in the face of the responses you are seeing? Vanguard's mission is to provide the same semantics in the face of protocol transformations.

FWIW, the behavior described thus far in this issue is still inline with [the gRPC wire-format spec](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md), wherein "trailers only" responses are allowed _but not required_.

> Same build can lead to different outcome depending on unknown to me yet conditions.

This is certainly unexpected. There shouldn't be anything non-deterministic in the implementation, unless there is something non-deterministic in the downstream server responses. Having said that, if the behavior is really non-deterministic (would be great to create a simple repro case, where one can clearly see different styles of responses with each invocation), that isn't _necessarily_ a bug, as long as the same semantics are conveyed in all response styles. But having a way to repro what you're seeing would go a long way to helping me understand the ostensible problem.

**Comment by rauanmayemir** [comment]
I appreciate you taking time answering me, I’ll come up with a detailed explanation later.
